marlowe-dashboard-client/src/Capability/Marlowe.purs

```diff
@@ -102,6 +107,32 @@ instance manageMarloweAppM :: ManageMarlowe AppM where
             , previousCompanionAppState: Nothing
             }
         pure $ createWalletDetails <$> ajaxCompanionAppId <*> ajaxMarloweAppId <*> ajaxAssets
+  restoreWallet options = do
+    mWalletInfo <- Wallet.restoreWallet options
+    case mWalletInfo of
+      Left err -> pure $ Left err
+      Right walletInfo -> do
+        let
+          walletId = view _walletId walletInfo
+        -- create the WalletCompanion and MarloweApp for this wallet
+        -- TODO: We had initial discussions for adding Variants for errors probably using Checked Exceptions.
+        -- Instead of the general ClientServerError we could encapsulate the type of error inside a context
+        -- (restoring the wallet, activating a wallet companion or marlowe app).
+        -- For the end users this wouldn't change too much their UX, but for production code it can be useful
+        -- to diagnose problems using a tool like sentry SCP-3171.
+        -- https://github.com/natefaubion/purescript-checked-exceptions
+        ajaxCompanionAppId <- lmap ClientServerError <$> Contract.activateContract WalletCompanion walletId
+        ajaxMarloweAppId <- lmap ClientServerError <$> Contract.activateContract MarloweApp walletId
+        let
```

Comment:
Cosmetics - do we avoid `puns` here for clarity?
 ```
createWalletDetails =
  { walletNickname: ""
  , companionIAppId: _
  , marloweAppId: _
  ...
  }
```

Reply:
not sure what you mean